src/main/java/io/reactivex/internal/operators/flowable/FlowableThrottleAndSample.java

```diff
@@ -0,0 +1,221 @@
+        final long windowDuration;
+        final TimeUnit unit;
+        final Scheduler.Worker worker;
+
+        final AtomicInteger timerInProgress;
+        final SequentialDisposable timer;
+
+        final AtomicLong requested = new AtomicLong();
+
+        Subscription subscription;
+
+        boolean isDone;
+
+        ThrottleAndSampleSubscriber(final Subscriber<? super T> actual, long windowDuration,
+                                    final TimeUnit unit, final Scheduler.Worker worker) {
+            this.actual = actual;
+            this.windowDuration = windowDuration;
+            this.unit = unit;
+            this.worker = worker;
+            this.timerInProgress = new AtomicInteger(0);
+            this.timer = new SequentialDisposable(Disposables.disposed());
+        }
+
+        @Override
+        public void onSubscribe(final Subscription subscription) {
+            if (SubscriptionHelper.validate(this.subscription, subscription)) {
+                this.subscription = subscription;
+                actual.onSubscribe(this);
+                subscription.request(Long.MAX_VALUE);
+            }
+        }
+
+        @Override
+        public void onNext(final T value) {
+            set(value);
+            if (timerInProgress.incrementAndGet() == 1) {
+                emit(getAndSet(null));
+                timer.update(worker.schedule(this, windowDuration, unit));
+            }
+        }
+
+        protected void emitFromRun() {
+            timer.update(Disposables.disposed());
+            if (timerInProgress.decrementAndGet() != 0) {
+                emit(getAndSet(null));
+                timer.replace(worker.schedule(this, windowDuration, unit));
+                timerInProgress.set(1);
+            }
+        }
+
+        @Override
+        public void onError(final Throwable t) {
+            if (isDone) {
+                RxJavaPlugins.onError(t);
+                return;
+            }
+            isDone = true;
+            cancelTimer();
+            actual.onError(t);
+        }
+
+        @Override
+        public void onComplete() {
+            if (isDone) {
+                return;
+            }
+            isDone = true;
+            cancelTimer();
+            complete();
+        }
+
+        @Override
+        public void request(final long n) {
+            if (SubscriptionHelper.validate(n)) {
+                BackpressureHelper.add(requested, n);
+            }
+        }
+
+        @Override
+        public void cancel() {
+            cancelTimer();
+            subscription.cancel();
+        }
+
+        private void cancelTimer() {
+            DisposableHelper.dispose(timer);
+            worker.dispose();
+        }
+
+        void emit(@NonNull final T value) {
+            long r = requested.get();
+            if (r != 0L) {
+                actual.onNext(value);
+                BackpressureHelper.produced(requested, 1);
+            } else {
+                cancel();
```

Comment:
`isDone` should be set here so that `onError`/`onComplete` doesn't get called again.